### DISCUSSION TITLE
Web mode

### DISCUSSION THREAD

**Discussion by dundee** [dundee/gdu#33] [Ideas]
Hello gdu contributors @fabaff @petasittek @egeoz @ouuan @c02y @lhl @kaihendry!
I am playing with an idea to add "web mode" into gdu. 
So it would start an http server and open a browser where you will see results of the scanning.

Do you think it would be a good addition to gdu?

You can check some very small basis of functionality using (go 1.16 is needed):

```bash
go get -u github.com/dundee/gdu/v5
$GOPATH/bin/gdu web /
```

**Comment by kaihendry** [comment]
I personally like the idea of a local Web mode, e.g. one of my projects for viewing images https://github.com/kaihendry/lk2

**Comment by ouuan** [comment]
I don't know how heavy the web mode is, so I'm wondering if it is a better idea to make the web mode a separate project. It can share codes with the cli version of gdu, or we can even have a libgdu.

**Reply by dundee** [reply]
Good idea! I will see how much will the web mode increase the size.

**Comment by c02y** [comment]
I personally don't think web-mode is necessary for such cli tool like gdu which is a disk usage analyzer in terminal, I use it in terminal, I don't like such kind of GUI app either, let alone such program in browser, like @ouuan said, if you like the web-mode idea, maybe you should start a new project.

**Comment by dundee** [comment]
Thanks for your answers. I will indeed develop the web mode as a separate project.

One more question for you: How do you like the CLI interface introduced in `v5`?
There are now commands (`help`, `version`, `interactive` and `disks`) instead of options. The usage now looks like this:

```
Usage:
  gdu [directory_to_scan] [flags]
  gdu [command]

Available Commands:
  disks       Show all mounted disks
  help        Help about any command
  interactive Run in interactive mode
  version     Print the version number of gdu

Flags:
  -h, --help                  help for gdu
  -i, --ignore-dirs strings   Absolute paths to ignore (separated by comma) (default [/proc,/dev,/sys,/run])
  -l, --log-file string       Path to a logfile (default "/dev/null")
  -c, --no-color              Do not use colorized output
  -x, --no-cross              Do not cross filesystem boundaries
  -p, --no-progress           Do not show progress in non-interactive mode
  -a, --show-apparent-size    Show apparent size
  ```

And there are shortcuts: `i` for `interactive` and `d` for `disks`.

It has one drawback - default mode is now non-interactive, but overall it looks a little bit more logical and sleek for me.

Few examples:
```
gdu /                      # same as current gdu -n /

gdu interactive /      # same as current gdu /
gdu i /                     # the same with shortcut

gdu disks               # same as current gdu -nd
gdu d                     # the same with shortcut

gdu interactive disks   # same as current gdu -d
gdu i d                         # the same with shortcuts
```

What do you think? Do you like more the new concept or the current interface?

**Reply by ouuan** [reply]
The new interface indicates that both `i` and `d` are subcommands, but the `gdu i d` usage shows that `i` is actually an option, not a subcommand. So I think it should be `-i` instead of `i`.

I personally prefer the current interface, but the new one is OK since I can use aliases in my zshrc.

**Reply by ouuan** [reply]
And it's better to always have a subcommand (`gdu something /` instead of `gdu /`), otherwise we need a way to distinguish subcommands and directory names.

**Reply by c02y** [reply]
Does gdu really need subcommands? Options are enough for me. Currently gdu is just a disk usage analyzer cli tool, now according to this thread, I don't know what the final gdu will be.

**Reply by dundee** [reply]
> The new interface indicates that both `i` and `d` are subcommands, but the `gdu i d` usage shows that `i` is actually an option, not a subcommand. So I think it should be `-i` instead of `i`.
> 
> I personally prefer the current interface, but the new one is OK since I can use aliases in my zshrc.

`d` is subcommand of gdu and also subsubcommand of `gdu i`.

I agree. It's not needed for gdu to have subcommands. I will introduce this interface in the new project with web interface, there it makes more sense. 

Thanks again for your replies!

**Comment by rwahyudi** [comment]
I think it would be better if the web mode is focusing on reading from generated reports rather than driving gdu in real time.  It would be mroe awesome if we could generate comparison between reports.

**Reply by rwahyudi** [reply]
Just replying to myself as I had a use case where user need the disk usage report. 
Script available from : https://github.com/rwahyudi/webdiskstat/

Example : https://htmlpreview.github.io/?https://github.com/rwahyudi/webdiskstat/blob/main/example/report.html

**Comment by rsheyd** [comment]
In case anyone is looking for a version of gdu with a web ui, I created a separate project here: https://github.com/rsheyd/gdu-web

**Reply by dundee** [reply]
Nice UI!

Few ideas to consider: 
* You have `github.com/roman/gdu-web` in go.mod but the project lives on different path in github.
* I would add two more target to the Makefile to simplify the setup:
```
install-dev-deps:
	cd ui && npm install

run:
	go run ./cmd/gdu-web
```

* The project claims to be fast and "The scanner is adapted from concepts in gdu", but it actually isn't. Gdu is built around the strong parallelism that Golang offers, but gdu-web is not using that. I would suggest to either try to use goroutines in your codebase (hard) or start using Gdu as a library, similarly as it's being used here (much easier): https://github.com/dundee/disk_usage_exporter/blob/master/exporter/exporter.go#L66

**Reply by rsheyd** [reply]
@dundee thank you for the feedback! I will definitely consider and follow up 🙏

**Reply by rsheyd** [reply]
@dundee 

Thank you for flagging the module path and the missing Makefile targets - I got those fixed.

Your point about the scanner is fair. I’ve clarified that the project was inspired by gdu but currently uses its own scanner implementation. I’m planning to investigate using gdu as a library in stages: first checking whether its APIs fit the project’s cancellation, progress, Windows filesystem, and output requirements, then building a small prototype if that looks promising.

You can see my analysis of the library feasibility here: https://github.com/rsheyd/gdu-web/blob/main/docs/roadmap.md#stage-1-library-feasibility-review--completed

I appreciate you taking the time to look through it and point me in a better direction!

**Reply by dundee** [reply]
Thank you, I will look into the feasibility review. Gdu might benefit from some of the suggestions.
